File: ninchatsdk/src/main/java/com/ninchat/sdk/activities/NinchatMediaActivity.java
package com.ninchat.sdk.activities;

import android.app.DownloadManager;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.net.Uri;
import android.os.Bundle;
import android.os.Environment;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.content.LocalBroadcastManager;
import android.view.View;
import android.widget.ImageView;
import android.widget.MediaController;
import android.widget.TextView;
import android.widget.VideoView;

import com.ninchat.sdk.GlideApp;
import com.ninchat.sdk.R;
import com.ninchat.sdk.models.NinchatFile;

public final class NinchatMediaActivity extends NinchatBaseActivity {

    protected static final String FILE_ID = "fileId";

    public static Intent getLaunchIntent(final Context context, final String fileId) {
        return new Intent(context, NinchatMediaActivity.class).putExtra(FILE_ID, fileId);
    }

    @Override
    protected int getLayoutRes() {
        return R.layout.activity_ninchat_media;
    }

    @Override
    protected boolean allowBackButton() {
        return true;
    }

    public void onToggleTopBar(final View view) {
        final View top = findViewById(R.id.ninchat_media_top);
        top.setVisibility(top.getVisibility() == View.GONE ? View.VISIBLE : View.GONE);
    }

    public void onClose(final View view) {
        finish();
    }

    public void onDownloadFile(final View view) {
        if (hasFileAccessPermissions()) {
            downloadFile();
        } else {
            requestFileAccessPermissions();
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        if (requestCode == STORAGE_PERMISSION_REQUEST_CODE) {
            if (hasFileAccessPermissions()) {
                downloadFile();
            } else {
                showError(R.id.ninchat_media_error, R.string.ninchat_chat_error_no_file_permissions);
            }
        }
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
    }

    private void downloadFile() {
        final NinchatFile file = sessionManager.getFile(fileId);
        final Uri uri = Uri.parse(file.getUrl());
        final DownloadManager.Request request = new DownloadManager.Request(uri);
        request.setTitle(file.getName());
        request.setDescription(file.getUrl());
        request.setDestinationInExternalPublicDir(Environment.DIRECTORY_PICTURES, file.getName());
        final DownloadManager downloadManager = (DownloadManager) getSystemService(Context.DOWNLOAD_SERVICE);
        downloadManager.enqueue(request);
        findViewById(R.id.ninchat_media_download).setVisibility(View.GONE);
    }

    private String fileId;

    private BroadcastReceiver fileDownloadedReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            final String action = intent.getAction();
            if (DownloadManager.ACTION_DOWNLOAD_COMPLETE.equals(action)) {
                findViewById(R.id.ninchat_media_download).setVisibility(sessionManager.getFile(fileId).isDownloaded() ? View.GONE : View.VISIBLE);
            }
        }
    };

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        fileId = getIntent().getStringExtra(FILE_ID);
        final NinchatFile file = sessionManager.getFile(fileId);
        findViewById(R.id.ninchat_media_download).setVisibility(file.isDownloaded() ? View.GONE : View.VISIBLE);
        if (file.isVideo()) {
            findViewById(R.id.ninchat_media_image).setVisibility(View.GONE);
            final VideoView video = findViewById(R.id.ninchat_media_video);
            MediaController mediaController = new MediaController(this);
            mediaController.setAnchorView(video);
            mediaController.setMediaPlayer(video);
            video.setVisibility(View.VISIBLE);
            video.setMediaController(mediaController);
            video.setVideoPath(file.getUrl());
            video.start();
        } else {
            final ImageView image = findViewById(R.id.ninchat_media_image);
            GlideApp.with(this)
                    .load(file.getUrl())
                    .into(image);
        }
        if (file.isDownloaded()) {
            findViewById(R.id.ninchat_media_download).setVisibility(View.GONE);
        }
        final TextView name = findViewById(R.id.ninchat_media_name);
        name.setText(file.getName());
        final LocalBroadcastManager localBroadcastManager = LocalBroadcastManager.getInstance(this);
        localBroadcastManager.registerReceiver(fileDownloadedReceiver, new IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE));
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        final LocalBroadcastManager localBroadcastManager = LocalBroadcastManager.getInstance(this);
        localBroadcastManager.unregisterReceiver(fileDownloadedReceiver);
    }
}

File: ninchatsdk/src/main/java/com/ninchat/sdk/ninchatmedia/view/NinchatMediaActivity.kt
package com.ninchat.sdk.ninchatmedia.view

import android.app.DownloadManager
import android.os.Build
import android.os.Bundle
import android.view.View
import android.view.animation.Animation
import android.view.animation.LinearInterpolator
import android.view.animation.RotateAnimation
import android.widget.ImageView
import com.ninchat.sdk.R
import com.ninchat.sdk.activities.NinchatBaseActivity
import com.ninchat.sdk.helper.glidewrapper.GlideWrapper
import com.ninchat.sdk.ninchatmedia.model.NinchatMediaModel
import com.ninchat.sdk.ninchatmedia.presenter.INinchatMediaCallback
import com.ninchat.sdk.ninchatmedia.presenter.INinchatMediaPresenter
import com.ninchat.sdk.ninchatmedia.presenter.NinchatMediaPresenter
import kotlinx.android.synthetic.main.activity_ninchat_media.*

class NinchatMediaActivity : NinchatBaseActivity(), INinchatMediaPresenter {
    override val layoutRes: Int
        get() = R.layout.activity_ninchat_media

    override fun allowBackButton(): Boolean {
        return true
    }

    // ninchat media presenter
    private val ninchatMediaPresenter = NinchatMediaPresenter(
            ninchatMediaModel = NinchatMediaModel(),
            callback = this,
            mContext = this
    )

    fun onToggleTopBar(view: View?) {
        ninchatMediaPresenter.toggleMediaTopBar(ninchat_media_top)
    }

    fun onClose(view: View?) {
        finish()
    }

    fun onDownloadFile(view: View?) {
        if (hasFileAccessPermissions()) {
            ninchatMediaPresenter
                    .downloadFile(ninchat_media_download, getSystemService(DOWNLOAD_SERVICE) as DownloadManager)
        } else {
            requestFileAccessPermissions()
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        if (requestCode == STORAGE_PERMISSION_REQUEST_CODE) {
            if (hasFileAccessPermissions()) {
                ninchatMediaPresenter
                        .downloadFile(ninchat_media_download, getSystemService(DOWNLOAD_SERVICE) as DownloadManager)
            } else {
                showError(R.id.ninchat_media_error, R.string.ninchat_chat_error_no_file_permissions)
            }
        }
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        ninchatMediaPresenter.updateFileId(intent = intent)
        ninchatMediaPresenter.ninchatMediaModel.getFile()?.let { ninchatFile ->
            if (ninchatFile.isVideo) {
                ninchatMediaPresenter.playVideo(ninchat_media_video, ninchatFile.url)
            } else {
                ninchat_loading_image_preview.visibility = View.VISIBLE
                ninchat_loading_image_preview_text.text = ninchatMediaPresenter.getLoadingText
                val spinner = animateSpinner()

                GlideWrapper.loadImage(this, ninchatFile.url, ninchat_media_image, object : INinchatMediaCallback {
                    override fun onLoadError() {
                        spinner?.cancel()
                        showError(R.id.ninchat_media_error, R.string.ninchat_chat_error_downloading_preview)
                    }

                    override fun onLoadSuccess() {
                        spinner?.cancel()
                        ninchat_loading_image_preview.visibility = View.GONE
                    }
                })
            }
            ninchat_media_name.text = ninchatFile.name
            ninchat_media_download.visibility = if (ninchatFile.isDownloaded || ninchatFile.isVideo) View.GONE else View.VISIBLE
        }
        ninchatMediaPresenter.subscribeBroadcaster()
    }

    override fun onDestroy() {
        ninchatMediaPresenter.unSubscribeBroadcaster()
        super.onDestroy()
    }

    override fun onDownloadComplete() {
        ninchat_media_download.visibility = if (ninchatMediaPresenter.ninchatMediaModel.isDownloaded()) View.GONE else View.VISIBLE
    }

    // helper class for test
    internal fun getMediaPresenter() = ninchatMediaPresenter


    private fun animateSpinner(): RotateAnimation? {
        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.M) {
            val spinnerAnimation = RotateAnimation(0f, 359f, Animation.RELATIVE_TO_SELF, 0.5f, Animation.RELATIVE_TO_SELF, 0.5f).apply {
                interpolator = LinearInterpolator()
                repeatCount = Animation.INFINITE
                duration = 3000
            }
            val spinner: ImageView = ninchat_loading_image_preview_spinner.apply {
                visibility = View.VISIBLE
                animation = spinnerAnimation
            }
            return spinnerAnimation
        }
        return null
    }

}